File: android/core/clients/src/main/kotlin/com/nexussoft/verbum/clients/NotificationClient.kt
package com.nexussoft.verbum.clients

import com.nexussoft.verbum.models.PassageReference
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emptyFlow

/**
 * Local notifications for the verse of the day. Mirrors the iOS `NotificationClient`:
 * features schedule a short plan of dated notifications (the text is resolved when the
 * plan is built, so delivery needs no network) and learn which verse was tapped.
 */
interface NotificationClient {
    suspend fun authorization(): NotificationAuthorization
    /** Asks once; the system remembers the answer. `true` when granted. */
    suspend fun requestAuthorization(): Boolean
    /** Replaces every pending verse notification with these. */
    suspend fun scheduleVerses(notifications: List<VerseNotification>)
    suspend fun cancelVerses()
    /** The verse of each notification the user taps, including the one that launched the app. */
    fun openedVerses(): Flow<PassageReference> = emptyFlow()
}

enum class NotificationAuthorization { NOT_DETERMINED, AUTHORIZED, DENIED }

/**
 * One dated local notification carrying a verse. Wall-clock fields, not an instant, so
 * "7:00" stays 7:00 across time-zone changes.
 */
data class VerseNotification(
    val year: Int,
    val month: Int,
    val day: Int,
    val hour: Int,
    val minute: Int,
    val title: String,
    val body: String,
    val reference: PassageReference,
) {
    /** Stable per day, so rescheduling replaces rather than duplicates. */
    val id: String get() = "dailyVerse.$year-$month-$day"
}

/** Preview / fixture implementation: grants everything, schedules nothing. */
object NoopNotificationClient : NotificationClient {
    override suspend fun authorization() = NotificationAuthorization.AUTHORIZED
    override suspend fun requestAuthorization() = true
    override suspend fun scheduleVerses(notifications: List<VerseNotification>) = Unit
    override suspend fun cancelVerses() = Unit
}

File: VerbumKit/Sources/Clients/NotificationClient.swift
import ComposableArchitecture
import Models

/// Local notifications for the verse of the day. Features schedule a short
/// plan of dated notifications (the text is resolved when the plan is built,
/// so delivery needs no network) and learn which verse was tapped.
@DependencyClient
public struct NotificationClient: Sendable {
    public var authorization: @Sendable () async -> NotificationAuthorization = { .notDetermined }
    /// Asks once; the system remembers the answer. `true` when granted.
    public var requestAuthorization: @Sendable () async throws -> Bool
    /// Replaces every pending verse notification with these.
    public var scheduleVerses: @Sendable (_ notifications: [VerseNotification]) async -> Void
    public var cancelVerses: @Sendable () async -> Void
    /// The verse of each notification the user taps, including the one that
    /// launched the app. Never finishes on its own.
    public var openedVerses: @Sendable () -> AsyncStream<PassageReference> = { .finished }
}

public enum NotificationAuthorization: Equatable, Sendable {
    case notDetermined, authorized, denied
}

/// One dated local notification carrying a verse. Wall-clock components, not a
/// `Date`, so "7:00" stays 7:00 across time-zone changes.
public struct VerseNotification: Equatable, Sendable {
    public let year: Int
    public let month: Int
    public let day: Int
    public let hour: Int
    public let minute: Int
    public let title: String
    public let body: String
    public let reference: PassageReference

    public init(year: Int, month: Int, day: Int, hour: Int, minute: Int, title: String, body: String, reference: PassageReference) {
        self.year = year
        self.month = month
        self.day = day
        self.hour = hour
        self.minute = minute
        self.title = title
        self.body = body
        self.reference = reference
    }

    /// Stable per day, so rescheduling replaces rather than duplicates.
    public var id: String { "dailyVerse.\(year)-\(month)-\(day)" }
}

extension NotificationClient: DependencyKey {
    public static let liveValue = NotificationClient.userNotifications
    public static let previewValue = NotificationClient(
        authorization: { .authorized },
        requestAuthorization: { true },
        scheduleVerses: { _ in },
        cancelVerses: {},
        openedVerses: { .finished }
    )
}

extension DependencyValues {
    public var notificationClient: NotificationClient {
        get { self[NotificationClient.self] }
        set { self[NotificationClient.self] = newValue }
    }
}
